docs/api_extra.rst

```diff
@@ -725,21 +748,52 @@ section <ndarrays>`.
       - `owner`: if provided, the array will hold a reference to this object
         until it is destructed.
 
-      - `strides` is optional; a value of ``nullptr`` implies C-style strides.
+      - `strides` is optional; a value of ``nullptr`` implies that the order
+        is computed automatically according to the ``strides`` parameter.
+        Note that the ``strides`` count elements, not bytes.
 
       - `dtype` describes the data type (floating point, signed/unsigned
         integer) and bit depth.
 
       - The `device_type` and `device_id` indicate the device and address
         space associated with the pointer `value`.
 
-   .. cpp:function:: ndarray(void * data, const std::initializer_list<size_t> shape, handle owner = nanobind::handle(), std::initializer_list<int64_t> strides = { }, dlpack::dtype dtype = nanobind::dtype<Scalar>(), int32_t device_type = device::cpu::value, int32_t device_id = 0)
+      - The `order` value denotes the coefficient order in memory and is only
+        relevant when `strides` is empty. Specify ``'C'`` for C-style or ``'F'``
+        for Fortran-style. When this parameter is not explicitly specified, the
+        implementation uses the ndarray template parameters and C-style order
+        as a fallback.
+
+      Both ``strides`` and ``shape`` will be copied by the constructor, hence
+      the targets of these pointers don't need to remain valid following the
+      constructor call.
+
+      The Python *global interpreter lock* (GIL) must be held when calling this
+      function.
+
+   .. cpp:function:: ndarray(VoidPtr data, const std::initializer_list<size_t> shape, handle owner, std::initializer_list<int64_t> strides = { }, dlpack::dtype dtype = nanobind::dtype<Scalar>(), int32_t device_type = DeviceType, int32_t device_id = 0, char order = Order)
 
       Alternative form of the above constructor, which accepts the ``shape``
       and ``strides`` arguments using a ``std::initializer_list``. It
       automatically infers the value of ``ndim`` based on the size of
       ``shape``.
 
+      Both ``strides`` and ``shape`` will be copied by the constructor, hence
+      the targets of these initializer lists don't need to remain valid
+      following the constructor call.
+
+      The Python *global interpreter lock* (GIL) must be held when calling this
+      function.
+
+   .. cpp:function:: ndarray(VoidPtr data)
```

Comment:
I really like the ``nb::ndarray<...>`` compile-time template interface. One can simply dump some constraints in any order. nanobind will enforce them in bindings, arrange them consistently in docstrings, and use defaults for missing attributes.

The ``nb::ndarray()`` run-time constructor is super-restrictive in comparison. You need to provide parameters in a certain order and can't skip irrelevant ones.  Ideally I would have a *single* template constructor that takes anything and detects what it is given, but this is tricky to fix now because several of these parameters have similar types. (e.g. integers or lists of integers).

Reply:
I gave it another attempt in 3f835139e89ae0b3b8caaed3586f54937f36c819. Still not as convenient as I would ideally like it, but this is probably the best that can be done with backward compatibility in mind.